src/topology/mod.rs

```diff
@@ -628,53 +647,70 @@ impl TopologyDescription {
-                              top_arc: Arc<RwLock<TopologyDescription>>,
-                              run_monitor: bool) {
+    fn update_rs_from_primary(
+        &mut self,
+        host: Host,
+        description: Arc<RwLock<ServerDescription>>,
+        client: Client,
+        top_arc: Arc<RwLock<TopologyDescription>>,
+        run_monitor: bool,
+    ) {
 
         if !self.servers.contains_key(&host) {
             return;
         }
 
+        let description_set_name = description.read().unwrap().set_name.clone();
+
         if self.set_name.is_empty() {
-            self.set_name = description.set_name.to_owned();
-        } else if self.set_name != description.set_name {
+            self.set_name = description_set_name;
+        } else if self.set_name != description_set_name {
             // Primary found, but it doesn't have the setName
             // provided by the user or previously discovered.
             self.servers.remove(&host);
             self.check_if_has_primary();
             return;
         }
 
-        if description.set_version.is_some() && description.election_id.is_some() {
-            if self.max_set_version.is_some() && self.max_election_id.is_some() &&
-               (self.max_set_version.unwrap() > description.set_version.unwrap() ||
-                (self.max_set_version.unwrap() == description.set_version.unwrap() &&
-                 self.max_election_id.as_ref().unwrap() >
-                 description.election_id.as_ref().unwrap())) {
-                // Stale primary
-                if let Some(server) = self.servers.get(&host) {
-                    {
-                        let mut server_description = server.description.write().unwrap();
-                        server_description.server_type = ServerType::Unknown;
-                        server_description.set_name = String::new();
-                        server_description.election_id = None;
+        let (description_set_version, description_election_id) = {
+            let description_guard = description.read().unwrap();
+            (
+                description_guard.set_version,
+                description_guard.election_id.clone(),
+            )
+        };
+
+        if let Some(set_version) = description_set_version {
+            if let Some(election_id) = description_election_id.clone() {
+                if let Some(max_set_version) = self.max_set_version {
+                    if let Some(max_election_id) = self.max_election_id.clone() {
+                        if max_set_version > set_version ||
+                            (max_set_version == set_version && max_election_id > election_id)
+                        {
+                            // Stale primary
+                            if let Some(server) = self.servers.get(&host) {
+                                {
+                                    let mut server_description =
+                                        server.description.write().unwrap();
```

Comment:
I belive this is where the error occured; since we were already passing a lock guard for a `ServerDescription` to this point, if the server we selected here is the same one, then trying to acquire a write lock causes deadlock.